src/cpp-common/include/cpp-common/IHashMap.h

```diff
@@ -0,0 +1,71 @@
+#ifndef PROPOLIS_IHASHMAP_H
+#define PROPOLIS_IHASHMAP_H
+
+#include <cstdint>
+#include <optional>
+#include <utility>
+
+/**
+ * @brief A templated hash map that can store a fixed number of items
+ * @tparam Key The key used to find elements in the hash map. Must be some sort of integer
+ * @tparam MappedType The type to store in the map. Can be any type
+ * @tparam maxSize The maximum number of items that can be stored in the map.
+ */
+template <typename Key, typename MappedType, uint16_t maxSize>
+class IHashMap {
+  public:
+    virtual ~IHashMap<Key, MappedType, maxSize>() = default;
+
+    /**
+     * @brief Insert an item into the hash map
+     * @return true if successfull, false if the map was full or key already present in map
+     */
+    virtual bool insert(const std::pair<Key, MappedType>& item) = 0;
+
+    /**
+     * @brief Obtain the copy of an item from the hash map based on a Key
+     * @param [in] key The key to get the item to get
+     * @param [out] item The reference to the item to store the value wanted
+     * @return true if succesfull, false if no item found based on Key
+     */
+    virtual bool get(Key key, MappedType& item) const = 0;
+
+    /**
+     * @brief Obtain the copy of an item from the hash map based on a Key
+     * @param key The key to get the item to get
+     * @return A reference to the mapped type if key was found in map, otherwise an empty optional
+     */
+    virtual std::optional<MappedType> at(Key key) const = 0;
```

Comment:
Should `at()` or maybe a new call return a reference to the entry in the hashmap. If we need to update the value without wanting to delete and then readd it, it would be useful

Reply:
You should return a std::optional<std::reference_wrapper<MappedType>>, if you use directly MappedType& you wont be able to return partial/abstract type. This is a limitation of std::optional, but std::reference_wrapper removes this limitation

Reply:
Will have to return a pointer for this, since an optional non-const reference is not allowed in c++